ksqldb-engine/src/main/java/io/confluent/ksql/engine/KsqlEngine.java

```diff
@@ -252,6 +258,39 @@ public static boolean isExecutableStatement(final Statement statement) {
         || statement instanceof Query;
   }
 
+  /**
+   * Resets a query internal KafkaStreams so it can be restarted.
+   * </p>
+   * When a KafkaStreams has been stopped, it cannot be started again. To allow a restart,
+   * the rest creates a new new KafkaStreams with the same topology and configs.
+   *
+   * @param queryId the queryId to reset.
+   */
+  public void resetQuery(final QueryId queryId) {
+    if (primaryContext.getPersistentQueries().containsKey(queryId)) {
+      final PersistentQueryMetadata query = primaryContext.getPersistentQueries().get(queryId);
+      if (query.getState() != State.NOT_RUNNING) {
+        throw new IllegalStateException(
+            String.format("Cannot reset query with application id: %s because is in %s state",
+                query.getQueryApplicationId(), query.getState()));
+      }
+
+      // Unregister the query to replace it with a new PersistentQueryMetadata
+      primaryContext.unregisterQuery(query);
+      // this.unregisterQuery() is called from withing primaryContext.unregisterQuery()
+
+      final Topology topology = query.getTopology();
```

Comment:
I initially encapsulated this logic there, but I had a design problem with it. The `QueryMetadata` accepts a `KafkaStreams` object which I think should be immutable. If I have a `restart` command, then I would need to reset the internal KafkaStreams by creating a new one on every restart. 

I was thinking of creating the KafkaStreams inside the QueryMetadata instead of passing it as a parameter, but that KafkaStreams is used by the MaterializationProvider, which is created before constructing the QueryMetadata. Btw, I now need to replace the KafkaStreams from the MaterializationProvider too (see `QueryExecutor.buildQuery`).

I need more refactoring on these classes.

Reply:
@rodesai Done. I encapsulate the logic in a restart() method now.